File: chat-uikit/src/main/java/io/agora/chat/uikit/manager/EaseNotificationMsgManager.java
package io.agora.chat.uikit.manager;

import android.text.TextUtils;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.Conversation;
import io.agora.chat.TextMessageBody;
import io.agora.chat.uikit.constants.EaseConstant;

public class EaseNotificationMsgManager {
    private static EaseNotificationMsgManager instance;

    private EaseNotificationMsgManager(){}

    public static EaseNotificationMsgManager getInstance() {
        if(instance == null) {
            synchronized (EaseNotificationMsgManager.class) {
                if(instance == null) {
                    instance = new EaseNotificationMsgManager();
                }
            }
        }
        return instance;
    }

    /**
     * Create notification message
     * @param message
     * @param ext
     * @return
     */
    public ChatMessage createMessage(String message, Map<String, Object> ext) {
        ChatMessage emMessage = ChatMessage.createReceiveMessage(ChatMessage.Type.TXT);
        emMessage.setFrom(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID);
        emMessage.setMsgId(UUID.randomUUID().toString());
        emMessage.setStatus(ChatMessage.Status.SUCCESS);
        emMessage.addBody(new TextMessageBody(message));
        if(ext != null && !ext.isEmpty()) {
            Iterator<String> iterator = ext.keySet().iterator();
            while (iterator.hasNext()) {
                String key = iterator.next();
                Object value = ext.get(key);
                putObject(emMessage, key, value);
            }
        }
        emMessage.setUnread(true);
        ChatClient.getInstance().chatManager().saveMessage(emMessage);
        return emMessage;
    }

    private void putObject(ChatMessage message, String key, Object value) {
        if(TextUtils.isEmpty(key)) {
            return;
        }
        if(value instanceof String) {
            message.setAttribute(key, (String) value);
        }else if(value instanceof Byte) {
            message.setAttribute(key, (Integer) value);
        }else if(value instanceof Character) {
            message.setAttribute(key, (Integer) value);
        }else if(value instanceof Short) {
            message.setAttribute(key, (Integer) value);
        }else if(value instanceof Integer) {
            message.setAttribute(key, (Integer) value);
        }else if(value instanceof Boolean) {
            message.setAttribute(key, (Boolean) value);
        }else if(value instanceof Long) {
            message.setAttribute(key, (Long) value);
        }else if(value instanceof Float) {
            JSONObject object = new JSONObject();
            try {
                object.put(key, value);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            message.setAttribute(key, object);
        }else if(value instanceof Double) {
            JSONObject object = new JSONObject();
            try {
                object.put(key, value);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            message.setAttribute(key, object);
        }else if(value instanceof JSONObject) {
            message.setAttribute(key, (JSONObject) value);
        }else if(value instanceof JSONArray) {
            message.setAttribute(key, (JSONArray) value);
        }else {
            JSONObject object = new JSONObject();
            try {
                object.put(key, value);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            message.setAttribute(key, object);
        }
    }

    /**
     * Create ext map
     * @return
     */
    public Map<String, Object> createMsgExt() {
        return new HashMap<>();
    }

    /**
     * Get latest message
     * @param con
     * @return
     */
    public ChatMessage getLastMessageByConversation(Conversation con) {
        if(con == null) {
            return null;
        }
        return con.getLastMessage();
    }

    /**
     * Get notification conversation
     * @return
     */
    public Conversation getConversation() {
        return getConversation(true);
    }

    /**
     * Get notification conversation
     * @param createIfNotExists
     * @return
     */
    public Conversation getConversation(boolean createIfNotExists) {
        return ChatClient.getInstance().chatManager().getConversation(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID
                , Conversation.ConversationType.Chat, createIfNotExists);
    }

    /**
     * Get all messages of notification
     * @return
     */
    public List<ChatMessage> getAllMessages() {
        return getConversation().getAllMessages();
    }

    /**
     * Check whether is a notification message
     * @param message
     * @return
     */
    public boolean isNotificationMessage(ChatMessage message) {
        return message.getType() == ChatMessage.Type.TXT
                && TextUtils.equals(message.getFrom(), EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID);
    }

    /**
     * Check whether is a notification conversation
     * @param conversation
     * @return
     */
    public boolean isNotificationConversation(Conversation conversation) {
        return conversation.getType() == Conversation.ConversationType.Chat
                && TextUtils.equals(conversation.conversationId(), EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID);
    }

    /**
     * Get the message content
     * @param message
     * @return
     */
    public String getMessageContent(ChatMessage message) {
        if(message.getBody() instanceof TextMessageBody) {
            return ((TextMessageBody)message.getBody()).getMessage();
        }
        return "";
    }

    /**
     * Update notification message
     * @param message
     * @return
     */
    public boolean updateMessage(ChatMessage message) {
        if(message == null || !isNotificationMessage(message)) {
            return false;
        }
        ChatClient.getInstance().chatManager().updateMessage(message);
        return true;
    }

    /**
     * Remove notification message
     * @param message
     * @return
     */
    public boolean removeMessage(ChatMessage message) {
        if(message == null || !isNotificationMessage(message)) {
            return false;
        }
        Conversation conversation = ChatClient.getInstance().chatManager().getConversation(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID);
        conversation.removeMessage(message.getMsgId());
        return true;
    }

    /**
     * Make all message in notification conversation as read
     */
    public void markAllMessagesAsRead() {
        Conversation conversation = ChatClient.getInstance().chatManager().getConversation(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID);
        conversation.markAllMessagesAsRead();
    }
}

File: ease-im-kit/src/main/kotlin/com/hyphenate/easeui/feature/invitation/helper/EaseNotificationMsgManager.kt
package com.hyphenate.easeui.feature.invitation.helper

import android.text.TextUtils
import com.hyphenate.easeui.EaseIM
import com.hyphenate.easeui.common.ChatClient
import com.hyphenate.easeui.common.ChatConversation
import com.hyphenate.easeui.common.ChatConversationType
import com.hyphenate.easeui.common.ChatMessage
import com.hyphenate.easeui.common.ChatMessageStatus
import com.hyphenate.easeui.common.ChatMessageType
import com.hyphenate.easeui.common.ChatTextMessageBody
import com.hyphenate.easeui.common.EaseConstant
import com.hyphenate.easeui.common.bus.EaseFlowBus
import com.hyphenate.easeui.common.extensions.mainScope
import com.hyphenate.easeui.model.EaseEvent
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.UUID


class EaseNotificationMsgManager {

    companion object {

        private var instance: EaseNotificationMsgManager? = null
        fun getInstance(): EaseNotificationMsgManager {
            if (instance == null) {
                synchronized(EaseNotificationMsgManager::class.java) {
                    if (instance == null) {
                        instance = EaseNotificationMsgManager()
                    }
                }
            }
            return instance!!
        }
    }

    fun createMessage(message: String?, ext: Map<String, Any>): ChatMessage {
        val emMessage: ChatMessage = ChatMessage.createReceiveMessage(ChatMessageType.TXT)
        emMessage.from = EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID
        emMessage.msgId = UUID.randomUUID().toString()
        emMessage.setStatus(ChatMessageStatus.SUCCESS)
        emMessage.addBody(ChatTextMessageBody(message))
        if (ext.isNotEmpty()) {
            val iterator = ext.keys.iterator()
            while (iterator.hasNext()) {
                val key = iterator.next()
                val value = ext[key]
                value?.let {
                    putObject(emMessage, key, it)
                }
            }
        }
        emMessage.isUnread = true
        ChatClient.getInstance().chatManager().saveMessage(emMessage)
        return emMessage
    }

    private fun putObject(message: ChatMessage, key: String, value: Any) {
        if (TextUtils.isEmpty(key)) {
            return
        }
        when (value) {
            is String -> {
                message.setAttribute(key, value)
            }

            is Byte -> {
                message.setAttribute(key, (value as Int))
            }

            is Char -> {
                message.setAttribute(key, (value as Int))
            }

            is Short -> {
                message.setAttribute(key, (value as Int))
            }

            is Int -> {
                message.setAttribute(key, value)
            }

            is Boolean -> {
                message.setAttribute(key, value)
            }

            is Long -> {
                message.setAttribute(key, value)
            }

            is Float -> {
                val `object` = JSONObject()
                try {
                    `object`.put(key, value)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
                message.setAttribute(key, `object`)
            }

            is Double -> {
                val `object` = JSONObject()
                try {
                    `object`.put(key, value)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
                message.setAttribute(key, `object`)
            }

            is JSONObject -> {
                message.setAttribute(key, value)
            }

            is JSONArray -> {
                message.setAttribute(key, value)
            }

            else -> {
                val `object` = JSONObject()
                try {
                    `object`.put(key, value)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
                message.setAttribute(key, `object`)
            }
        }
    }

    fun createMsgExt(): MutableMap<String, Any> {
        return HashMap()
    }

    /**
     * Get latest message
     * @param con
     * @return
     */
    fun getLastMessageByConversation(con: ChatConversation?): ChatMessage? {
        return con?.lastMessage
    }


    /**
     * Get notification conversation
     * @return
     */
    fun getConversation():ChatConversation{
        return getSystemConversation(true)
    }

    /**
     * Get notification conversation
     * @param createIfNotExists
     * @return
     */
    fun getSystemConversation(createIfNotExists:Boolean):ChatConversation{
        return ChatClient.getInstance().chatManager().getConversation(
            EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID,
            ChatConversationType.Chat,
            createIfNotExists
        )
    }

    /**
     * Get all messages of notification
     * @return
     */
    fun getAllNotifyMessage():List<ChatMessage>{
        return getConversation().allMessages
    }

    /**
     * load more message of notification
     * @return
     */
    fun loadMoreMessage(startMsgId:String?="",limit:Int):List<ChatMessage>{
        return getConversation().loadMoreMsgFromDB(startMsgId,limit)
    }


    /**
     * Check whether is a notification message
     * @param message
     * @return
     */
    fun isNotificationMessage(message: ChatMessage): Boolean {
        return (message.type === ChatMessageType.TXT
                && TextUtils.equals(message.from, EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID))
    }


    /**
     * Check whether is a notification conversation
     * @param conversation
     * @return
     */
    fun isNotificationConversation(conversation: ChatConversation): Boolean {
        return (conversation.type === ChatConversationType.Chat
            && TextUtils.equals(conversation.conversationId(), EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID)
        )
    }


    /**
     * Get the message content
     * @param message
     * @return
     */
    fun getMessageContent(message: ChatMessage): String? {
        return if (message.body is ChatTextMessageBody) {
            (message.body as ChatTextMessageBody).message
        } else ""
    }

    /**
     * Update notification message
     * @param message
     * @return
     */
    fun updateMessage(message: ChatMessage?): Boolean {
        if (message == null || !isNotificationMessage(message)) {
            return false
        }
        ChatClient.getInstance().chatManager().updateMessage(message)
        return true
    }

    /**
     * Remove notification message
     * @param message
     * @return
     */
    fun removeMessage(message: ChatMessage?): Boolean {
        if (message == null || !isNotificationMessage(message)) {
            return false
        }
        val conversation: ChatConversation = ChatClient.getInstance().chatManager()
            .getConversation(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID)
        conversation.removeMessage(message.msgId)
        return true
    }

    /**
     * Make all message in notification conversation as read
     */
    fun markAllMessagesAsRead() {
        val conversation: ChatConversation = ChatClient.getInstance().chatManager()
            .getConversation(EaseConstant.DEFAULT_SYSTEM_MESSAGE_ID)
        conversation.markAllMessagesAsRead()

        val context = EaseIM.getContext()
        context?.let {
            it.mainScope().launch {
                EaseFlowBus.with<EaseEvent>(EaseEvent.EVENT.UPDATE.name).post(this,
                    EaseEvent(EaseEvent.EVENT.UPDATE.name, EaseEvent.TYPE.NOTIFY)
                )
            }
        }
    }
}